documentation/pattern-match-avoid-duplication.md

```diff
@@ -0,0 +1,266 @@
+containing a list of `CheckedMatchCase`s. Each `CheckedMatchCase` exactly one
+pattern and one body that uses the bindings declared within said pattern.
+ 
+The problem with this solution is that it effectively duplicates the body as
+many times as patterns exist in the same group. This made compiling painfully
+slow, but the overhead on typechecking the body was not the main issue: Codegen
+would receive the bodies already duplicated, severely limiting what can be done
+at late codegen stage. Take into account that the codegen stage has the most
+complete version of the program that Jakt can have, since it needs all the
+concrete types that are used in order to generate working C++. In #1597 we can
+see that a ~300 line Jakt function (`typecheck_binary_operation`) is compiled
+to ~5500 lines of C++ code! That happened because the bodies for each case are
+large, and the bodies were duplicated many times.
+
+To ensure that similar functions to `typecheck_binary_operation` do not make
+the C++ compiler sit through 18x the amount of Jakt code, we'll add a second
+constraint:
+
+### Constraint 2: Multiple patterns, same binding types, only one body
+
+Take the following samples:
+
+#### Constraint 2.1: Multiple patterns with bindings
+```jakt
+enum Foo {
+    Bar(i32)
+    Baz
+}
+
+fn get_foo() -> Foo { ... }
+
+fn main() {
+    match get_foo() {
+        Bar(x) | Baz default(x = 1i32) {
+            // ... long code using `x`
+        }
+    }
+}
+```
+
+#### Constraint 2.2: Multiple patterns with no bindings
+```jakt
+enum Answer {
+    Is42
+    IsNot42(actual_value: u32)
+    IsNot42AndNoValue
+}
+
+fn get_answer() -> Answer { ... }
+
+fn main() {
+    match get_answer() {
+        Is42 => {
+            // long body 1
+        }
+        IsNot42 | IsNot42AndNoValue => {
+            // long body 2
+        }
+    }
+}
+```
+
+In both cases, we want Jakt to ideally only generate the long bodies once.
+
+## A partial improvement (#1597)
+
+PR #1597 addressed Constraint 2.1 by keeping the shared relationship between
+the patterns inside a pattern group and the body they jump to. This enables
+codegen to find pattern groups that have no bindings on them and easily group
+them by concatenating `case` labels or concatenating boolean expressions with
+`||`.
+
+Now each `CheckedMatchCase` contains a list of `CheckedMatchPattern` and one
+body that uses the bindings declared within said pattern. The implementation
+typechecks all bodies like the last version, but only selects the last checked
+body.
+
+The problem with this approach is that now the typechecker portion of this
+solution does not respect Constraint 1, since the last checked body is not
+enough to represent the case if bindings from the same parsed pattern group can
+have different binding types.
+
+This makes `tests/codegen/match_enum_same_body_different_impl.jakt` to fail
+codegen, since it calls `maybe_templated<JaktArray<i32>>` on all cases, when
```

Comment:
```suggestion
This causes `tests/codegen/match_enum_same_body_different_impl.jakt` to fail,
since it codegens a call to `maybe_templated<JaktArray<i32>>` on all cases, when
```